lvsfunc/deblock.py

```diff
@@ -258,41 +252,137 @@ def _get_strength_clip(clip: vs.VideoNode, strength: SupportsFloat) -> vs.VideoN
 
-    no_dpir_zones = []
+    no_dpir_zones = list[Range]()
 
     if zones:
+        cache_strength_clips = dict[float, vs.VideoNode]()
+
+        dpir_zones = dict[int | Tuple[int | None, int | None], vs.VideoNode]()
+
         for ranges, zstr in zones:
-            if zstr:
-                strength_clip = replace_ranges(
-                    strength_clip, zstr if isinstance(zstr, vs.VideoNode)
-                    else _get_strength_clip(clip_rgb, zstr), ranges
-                )
-            else:
+            if not zstr:
                 if isinstance(ranges, List):
                     no_dpir_zones.extend(ranges)
                 else:
                     no_dpir_zones.append(ranges)
 
-    run_dpir = DPIR(clip_rgb, strength_clip, **dpir_args)
+                continue
+
+            if isinstance(zstr, vs.VideoNode):
+                rstr_clip = zstr
+            else:
+                zstr = float(zstr)
+
+                if zstr not in cache_strength_clips:
+                    cache_strength_clips[zstr] = _get_strength_clip(clip_rgb, zstr)
+
+                rstr_clip = cache_strength_clips[zstr]
+
+            lranges = ranges if isinstance(ranges, List) else [ranges]
+
+            for rrange in lranges:
+                if rrange:
+                    dpir_zones[rrange] = rstr_clip
+
+        if len(dpir_zones) <= 2:
+            for rrange, sclip in dpir_zones.items():
+                zoned_strength_clip = replace_ranges(strength_clip, sclip, rrange)
+        else:
+            dpir_ranges_zones = {
+                range(*(
+                    (r, r + 1) if isinstance(r, int) else (r[0] or 0, r[1] + 1 if r[1] else clip.num_frames)
+                )): sclip for r, sclip in dpir_zones.items()
+            }
+
+            dpir_ranges_zones = {k: dpir_ranges_zones[k] for k in sorted(dpir_ranges_zones, key=lambda x: x.start)}
+            dpir_ranges_keys = list(dpir_ranges_zones.keys())
+
+            def _select_sclip(n: int) -> vs.VideoNode:
+                nonlocal dpir_ranges_zones, dpir_ranges_keys
+
+                for i, ranges in enumerate(dpir_ranges_keys):
+                    if n in ranges:
+                        if i > 0:
+                            dpir_ranges_keys = dpir_ranges_keys[i:] + dpir_ranges_keys[:i]
+                        return dpir_ranges_zones[ranges]
+
+                return strength_clip
+
+            zoned_strength_clip = strength_clip.std.FrameEval(_select_sclip)
+    else:
+        zoned_strength_clip = strength_clip
+
+    if backend is None:
+        if None in {cuda, fp16}:
+            try:
+                info = cast(Dict[str, int], core.trt.DeviceProperties(device_id))
+
+                fp16_available = info['major'] >= 7
+                trt_available = True
+            except BaseException:
+                fp16_available = False
+                trt_available = False
+
+        if cuda is None:
+            cuda = 'trt' if trt_available else _check_has_nvidia()
```

Comment:
I can't find a way after a very quick google, but verifying that the user has a suitable GPU for trt might be preferable over just automatically enabling it otherwise. For example for me it's slightly worse iirc.

Reply:
I'm very sure there is a way to get if trt will be faster based on the core.trt.DeviceProperties output
Output on my GPU:
```py
{
    'access_policy_max_window_size': 0, 'async_engine_count': 1, 'can_map_host_memory': 1,
    'can_use_host_pointer_for_registered_mem': 0, 'clock_rate': 1835000, 'compute_mode': 0,
    'compute_preemption_supported': 1, 'conccurrent_managed_access': 0, 'concurrent_kernels': 1,
    'cooperative_launch': 1, 'cooperative_multi_device_launch': 0, 'device_overlap': 1,
    'direct_managed_mem_access_from_host': 0, 'driver_version': 11060, 'ecc_enabled': 0,
    'global_l1_cache_supported': 1, 'host_native_atomic_supported': 0, 'integrated': 0,
    'is_multi_gpu_board': 0, 'kernel_exec_timeout_enabled': 1, 'l2_cache_size': 2097152,
    'local_l1_cache_supported': 1, 'major': 6, 'managed_memory': 1, 'max_blocks_per_multi_processor': 32,
    'max_threads_per_block': 1024, 'max_threads_per_multiprocessor': 2048, 'mem_pitch': 2147483647,
    'memory_bus_width': 256, 'memory_clock_rate': 5505000, 'minor': 1, 'multi_gpu_board_group_id': 0,
    'multi_processor_count': 20, 'name': b'NVIDIA GeForce GTX 1080', 'pageable_memory_access': 0,
    'pageable_memory_access_uses_host_page_tables': 0, 'pci_bus_id': 45, 'pci_device_id': 0,
    'pci_domain_id': 0, 'persisting_l2_cache_max_size': 0, 'regs_per_block': 65536,
    'regs_per_multiprocessor': 65536, 'reserved_shared_mem_per_block': 0, 'shared_mem_per_block_optin': 49152,
    'shared_mem_per_multiprocessor': 98304, 'shared_memory_per_block': 49152, 'single_to_double_precision_perf_ratio': 32,
    'stream_priorities_supported': 1, 'tcc_driver': 0, 'texture_alignment': 512, 'texture_pitch_alignment': 32,
    'total_const_mem': 65536, 'total_global_memory': 8589606912, 'unified_addressing': 1,
    'uuid': [-48, 5, 119, 118, -102, -79, -85, -110, -13, -15, 117, -52, -15, 2, -36, 51], 'warp_size': 32
}
```

Reply:
It'll be a matter of figuring out *how* then. Got any leads?

Reply:
After some research, i can't really say this is even a good idea anymore.

Reply:
Let's scratch it then